server/plugin.go

```diff
@@ -84,102 +83,96 @@ func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Req
 		return
 	}
 
+	var err error
 	path := r.URL.Path
 	switch path {
 	case "/api/v1/product_type":
-		p.handleProductType(w, r)
+		err = p.handleProductType(w, r)
 	case "/api/v1/create_meeting_in_server_version":
 		p.handleCreateMeetingInServerVersion(w, r)
 	case "/api/v1/client_id":
-		p.handleClientID(w, r)
+		err = p.handleClientID(w, r)
 	case "/api/v1/auth":
-		p.handleAuthorizeInADD(w, r)
+		err = p.handleAuthorizeInADD(w, r)
 	case "/api/v1/auth_redirect":
-		p.completeAuthorizeInADD(w, r)
+		err = p.completeAuthorizeInADD(w, r)
 	case "/api/v1/register_meeting_from_online_version":
-		p.handleRegisterMeetingFromOnlineVersion(w, r)
+		err = p.handleRegisterMeetingFromOnlineVersion(w, r)
 	case "/api/v1/assets/profile.png":
 		p.handleProfileImage(w, r)
 	default:
 		http.NotFound(w, r)
 	}
+
+	if err != nil {
+		p.API.LogWarn(err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+	}
 }
 
-func (p *Plugin) handleAuthorizeInADD(w http.ResponseWriter, r *http.Request) {
+func (p *Plugin) handleAuthorizeInADD(w http.ResponseWriter, r *http.Request) error {
 
 	userID := r.URL.Query().Get("mattermost_user_id")
 
 	if userID == "" {
-		fmt.Println("Not authorized")
-		http.Error(w, "Not authorized", http.StatusUnauthorized)
-		return
+		return errors.New("cannot authorize in ADD. Missing 'mattermost_user_id' param")
```

Comment:
I have mixed feelings with this. You are returning the error, and letting `ServeHTTP` to write the error on the response, but `ServeHTTP` does not the nature of the error, so it will always return `StatusBadRequest`. I know this simplify many things, but I think it is better if each command still writes into the response. That way the HTTP Status code is the correct one. `ServeHTTP` can still do the logging.

Reply:
2/5 Each method could also return an http status code which get used my the caller

Reply:
Of course, you're right. I've changed the functions to return http status codes too :+1: